x/data/server/lookup/table.go

```diff
@@ -0,0 +1,159 @@
+	return NewTableWithOptions(TableOptions{
+		Prefix: prefix,
+	})
+}
+
+// NewTableWithOptions creates a Table with custom options. Most users should just use NewTable
+// with the default values.
+func NewTableWithOptions(options TableOptions) (Table, error) {
+	prefixLen := len(options.Prefix)
+	minLength := options.MinLength
+	if minLength == 0 {
+		minLength = 4
+	}
+
+	newHash := options.NewHash
+	if newHash == nil {
+		newHash = func() hash.Hash {
+			return fnv.New64a()
+		}
+	}
+
+	hashLen := len(newHash().Sum(nil))
+	if minLength > hashLen {
+		return nil, fmt.Errorf("option MinLength %d is greater than hash length %d", minLength, hashLen)
+	}
+
+	bufLen := prefixLen + hashLen + binary.MaxVarintLen64
+
+	return table{
+		minLen:    minLength,
+		bufLen:    bufLen,
+		newHash:   newHash,
+		prefix:    options.Prefix,
+		prefixLen: prefixLen,
+		hashLen:   hashLen,
+	}, nil
+}
+
+// TableOptions is used to specify custom table options and should only be used by advanced users.
+type TableOptions struct {
+	// NewHash is a function which returns a new hash.Hash instance.
+	NewHash func() hash.Hash
+
+	// MinLength is the minimum number of hash bytes that will be used to create a lookup identifier.
+	MinLength int
+
+	// Prefix is an optional prefix to be pre-pended to all KVStore keys.
+	Prefix []byte
+}
+
+type table struct {
+	minLen    int
+	bufLen    int
+	newHash   func() hash.Hash
+	prefix    []byte
+	prefixLen int
+	hashLen   int
+}
+
+// KVSTore is the interface for key-value stores that Tables operate on.
+type KVStore interface {
+	// Get returns nil iff key doesn't exist. Panics on nil key.
+	Get(key []byte) []byte
+
+	// Set sets the key. Panics on nil key or value.
+	Set(key, value []byte)
+}
+
+func (t table) GetValue(store KVStore, id []byte) []byte {
+	buf := make([]byte, t.prefixLen+len(id))
+	copy(buf, t.prefix)
+	copy(buf[t.prefixLen:], id)
+	return store.Get(id)
+}
+
+func (t table) GetOrCreateID(store KVStore, value []byte) []byte {
+	id, _ := t.getOrCreateID(store, value)
+	return id
+}
+
+func (t table) getOrCreateID(store KVStore, value []byte) (id []byte, numCollisions int) {
+	hasher := t.newHash()
+	_, err := hasher.Write(value)
+	if err != nil {
+		panic(err)
+	}
+	hashBz := hasher.Sum(nil)
+
+	id = make([]byte, 0, t.bufLen)
+	id = append(id, t.prefix...)
+
+	// take the first i bytes of hashBz starting with t.minLen and increasing
+	// in cases where there are collisions
+	for i := t.minLen; i <= t.hashLen; i++ {
+		id = append(id[t.prefixLen:], hashBz[:i]...)
+		if tryID(store, id, value) {
+			return id, i - t.minLen
+		}
+		id = id[:t.prefixLen]
+	}
+
+	// Deal with collisions by appending a varint disambiguation value.
+	// Such collisions are almost impossible with good settings, but can
+	// happen with a sub-optimal hash function.
+	var i uint64 = 0
+	preLen := t.prefixLen + t.hashLen
+	for {
+		id = id[:t.bufLen]
+		n := binary.PutUvarint(id[preLen:], i)
+		id = id[:preLen+n]
```

Comment:
```suggestion
		n := binary.PutUvarint(id[preLen:], i)
		copy(id[t.bufLen:], n)
```

Reply:
Otherwise you need to use `append(id, n...)` instead of `id = id[:preLen+n]`

Reply:
hmm... that doesn't quite work...

Reply:
I left it as is because this wasn't working